assertsprocessor/sampler.go

```diff
@@ -0,0 +1,195 @@
+package assertsprocessor
+
+import (
+	"context"
+	"github.com/tilinna/clock"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+	"math"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+)
+
+type traceQueues struct {
+	slowQueue  *TraceQueue
+	errorQueue *TraceQueue
+}
+
+func (tQ *traceQueues) errorTraceCount() int {
+	return len(tQ.errorQueue.priorityQueue)
+}
+
+func (tQ *traceQueues) slowTraceCount() int {
+	return len(tQ.slowQueue.priorityQueue)
+}
+
+type sampler struct {
+	logger               *zap.Logger
+	config               *Config
+	thresholdHelper      *thresholdHelper
+	topTracesMap         *sync.Map
+	healthySamplingState *sync.Map
+	traceFlushTicker     *clock.Ticker
+	nextConsumer         consumer.Traces
+	stop                 chan bool
+	requestRegexps       *map[string]*regexp.Regexp
+}
+
+type traceSummary struct {
+	hasError        bool
+	isSlow          bool
+	slowestRootSpan ptrace.Span
+	requestKey      RequestKey
+	latency         float64
+}
+
+func (s *sampler) sampleTrace(ctx context.Context,
+	trace ptrace.Traces, traceId string, spanSet *resourceSpanGroup) {
+	summary := s.getSummary(traceId, spanSet)
+	item := Item{
+		trace:   &trace,
+		ctx:     &ctx,
+		latency: summary.latency,
+	}
+	if summary.hasError || summary.isSlow {
+		// Get the trace queue for the entity and request
+		pq := s.getTraceQueues(summary.requestKey)
+		if summary.hasError {
+			s.logger.Info("Capturing error trace",
+				zap.String("traceId", summary.slowestRootSpan.TraceID().String()),
+				zap.Float64("latency", summary.latency))
+			pq.errorQueue.push(&item)
+		} else {
+			s.logger.Info("Capturing slow trace",
+				zap.String("traceId", summary.slowestRootSpan.TraceID().String()),
+				zap.Float64("latency", summary.latency))
+			pq.slowQueue.push(&item)
```

Comment:
We should have a check to ensure that the total number of traces across all priority queues does not exceed a max value. A trace should be added to a queue only if it's below the overall limit

Reply:
> We should have a check to ensure that the total number of traces across all priority queues does not exceed a max value. A trace should be added to a queue only if it's below the overall limit

The limit you are proposing is being enforced just before flushing the traces to the server, but it is only being done for the slow traces. The error trace is only subject to a 5/min/request limit. This is because using error trace's latency as a priority could lead to error traces being dropped. I think this needs some more thought.